sycl/doc/extensions/proposed/sycl_ext_oneapi_graph_fusion.asciidoc

```diff
@@ -0,0 +1,644 @@
+runtime**. This requires the extension of the runtime with some sort of JIT
+compiler to allow for the fusion of kernel functions at runtime.
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine the
+macro `SYCL_EXT_ONEAPI_GRAPH_FUSION` to one of the values defined in the
+table below.  Applications can test for the existence of this macro to determine
+if the implementation supports this feature, or applications can test the
+macro's value to determine which of the extension's features the implementation
+supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|Initial version of this extension.
+|===
+
+=== API modifications
+
+==== Properties
+
+===== Graph Fusion Property
+
+The API for `command_graph<graph_state::modifiable>::finalize()` includes a
+`property_list` parameter. The following property, defined by this extension,
+can be added to the property list to indicate that the kernels in the
+command-graph should be fused. 
+
+```c++
+sycl::ext::oneapi::experimental::property::command_graph::perform_fusion 
+```
+
+The property is not prescriptive. Implementations are free to not perform fusion
+if it is not possible (see below section <<limitations, Limitations>>), fusion is not
+supported by the implementation, or the implementation decides not to perform
+fusion for other reasons. It is not an error if an implementation does not
+perform fusion even though the property is passed. 
+
+Implementations can provide a diagnostic message in case fusion was not
+performed through an implementation-specified mechanism, but are not required to
+do so.
+
+===== Barrier property
+
+The following property can be added to the `property_list` of the
+`command_graph<graph_state::modifiable>::finalize()` API.
+
+```c++
+sycl::ext::oneapi::experimental::property::command_graph::no_barriers
+```
+
+If the property list contains this property, no barriers are introduced between
+kernels in the fused kernel (see below section on synchronization on kernels). 
+
+The property only takes effect if the
+`sycl::ext::oneapi::experimental::property::command_graph::perform_fusion`
+property is also part of the `property_list` of the same invocation of
+`command_graph<...>::finalize()`. 
+
+===== Local internalization property
+
+The following property can be passed to three different APIs, namely:
+
+* The `accessor` constructor, giving a more granular control.
+* The `buffer` constructor, in which case all the `accessors` derived from 
+this buffer will inherit this property (unless overridden).
+* The `property_list` parameter of `sycl::malloc_device()`,
+`sycl::aligned_alloc_device()`, `sycl::malloc_shared()`, or
+`sycl::aligned_alloc_shared()` to apply the property to an USM pointer.
+
+```c++ 
+sycl::ext::oneapi::experimental::property::promote_local 
```

Comment:
I think the name `promote` here (and in `promote_private`) is confusing.

I understand this is what is happening from the compiler's perspective, but it reads like the developer is prescribing promotion must happen.  In my mind, that's inconsistent with the fact that the compiler is allowed to ignore this property.

Since promotion isn't guaranteed, my recommendation here would be to find a way to express this which is more descriptive.  For example, a name like `work_group_local` or `access_scope<memory_scope::work_group>` would convey that a developer can use these properties to declare certain guarantees that may be used for optimization, without the developer having to prescribe which optimization is expected to happen.

Reply:
Oh, interesting.  I'd missed that.

Would it make sense to split that out into two orthogonal properties: one that expresses there is no sharing happening beyond a certain scope, and another that expresses there is no need to write the results back to global memory?

I understand that the current implementation would require both properties to be present to have any effect, but hypothetically speaking, couldn't it be useful to just see one of them?  I might be thinking about this the wrong way, but it seems like knowing that there's no sharing would be sufficient to cache/re-use the results in registers or work-group local memory during execution of some of the fused kernels, as long as the results were written out to global memory before they're needed.  Similarly, knowing that a buffer is only intended to be used by a specific kernel would allow the memory for those buffers to be re-used as scratchpads.

Reply:
Interesting thought!

> it seems like knowing that there's no sharing would be sufficient to cache/re-use the results in registers or work-group local memory during execution of some of the fused kernels, as long as the results were written out to global memory before they're needed

Yes, the current implementation does not support that, but the knowledge that no sharing occurs would be sufficient for internalization. The compiler would need to determine the stores which need to be kept to have the final result in the buffer again (or simply keep all stores), but that is only a limitation of the current implementation.

> Similarly, knowing that a buffer is only intended to be used by a specific kernel would allow the memory for those buffers to be re-used as scratchpads.

Can you elaborate a bit on what optimization the use as scratchpads would enable? So far, we had thought of not allocating the buffer at all as a potential optimization, but that requires successful internalization, i.e., the no-sharing property, too.

The main reason to have a combined property so far was to reduce the number of properties a user would have to specify to enable internalization. However, if each of the two aspects separately could already enable some optimization, it might indeed make sense to split the properties.

Reply:
> Can you elaborate a bit on what optimization the use as scratchpads would enable? So far, we had thought of not allocating the buffer at all as a potential optimization, but that requires successful internalization, i.e., the no-sharing property, too.

Disclaimer: I haven't fully thought this through!

I'm imagining a case something like this:

```c++
buffer<int> b1{range{dataSize}};
buffer<int> b2{range{dataSize}};
buffer<int> b3{range{dataSize}};

ext::oneapi::experimental::command_graph graph{
    q.get_context(), q.get_device()};

graph.begin_recording(q);

q.submit([&](handler &cgh) {
  auto a1 = b1.get_access(cgh);
  auto a2 = b2.get_access(cgh, sycl::ext::oneapi::experimental::property::scratch);
  cgh.parallel_for<K1>(dataSize, ...);
});

q.submit([&](handler &cgh) {
  auto a1 = b1.get_access(cgh);
  auto a2 = b3.get_access(cgh, sycl::ext::oneapi::experimental::property::scratch);
  cgh.parallel_for<K2>(dataSize, ...);
});

graph.end_recording();
```

`K1` is going to use `b2` to communicate some partial results during execution of the kernel, and `K2` is going to use `b3` for the same.  The values of `b2` and `b3` are never checked after the kernel executes.

Without some sort of decoration, it seems like a runtime would have no choice but to allocate storage for both `b2` and `b3`, just in case the values are used later.  With some sort of decoration, it would be legal to use the same storage for `b2` and `b3` within the graph.  I don't think this would always be true after fusion, but it would be true in some cases.

Somebody writing the program as I did above would probably have used a single buffer instead of `b2` and `b3`, but it's not hard to imagine a case like this arising where the kernels are defined and provided by different developers (e.g. in different libraries).

I seem to recall there was some discussion of supporting this sort of optimization in the original graph proposal, but I honestly don't remember where it landed.  If there's already a property with the semantics that I discussed above, maybe we could just use that + the scope-narrowing property.

Reply:
I like this direction a lot.

> My initial idea no_write_back isn't very suitable, as it would be easily confused with set_write_back(false), which is different as it disables write-back from device memory into host memory. Suggestions for the name are welcome 😉

Best I can come up with is `no_keep` or `no_save`.  I thought about `discard`, but I fear that would be too confusing given that `discard_read` and `discard_write` used to mean "discard the initial results".

Reply:
I split the properties' semantics as discussed. Access pattern/behavior is now expressed through the `access_scope` property template. Not initializing the internalized memory with values from the buffer and not writing back to the buffer are now expressed through the existing `no_init` property and the new `fusion_internal_memory` property, respectively. 

A few questions are still open and I would appreciate your opinion on those:
* Should we prohibit specialization of the `access_scope` property template with other values of the `memory_scope` enum beyond the two cases currently defined? Or should implementations be free to allow additional specializations?
* For the `no_init` property, the SYCL specification defines a list of places where it can appear (table 52). For our case, we need to allow it in additional places, do you see any issue with that?
* The name of the property for no write-back (`fusion_internal_memory`) is the best I could come up with so far, but can be changed if we have better ideas.